### PR TITLE
Allow matching empty frames in quality checks

### PR DESCRIPTION
**Pull request by zhiltsov-max**
### Motivation and context

Depends on #8634

Added a quality check option to consider frames matching, if both GT and job annotations have no annotations on a frame. This affects quality metrics and total counts in reports, but confusion matrices stay unchanged. This allows to use both positive and negative validation frames.

### How has this been tested?

Unit tests

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have created a changelog fragment 
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [ ] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/cvat-ai/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/cvat-ai/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/cvat-ai/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/cvat-ai/cvat/tree/develop/cvat-ui#versioning))

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/cvat-ai/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.

## Release Notes

- **New Features**
	- Introduced a quality setting for comparing point groups without bounding boxes.
	- Added an option to consider empty frames as matching in quality checks.
	- Enhanced quality settings form with new options for `matchEmptyFrames` and `useBboxSizeForPoints`.

- **Bug Fixes**
	- Corrected property name from `peoject_id` to `project_id` in the API quality reports filter.

- **Documentation**
	- Updated API schemas to include new quality settings properties. 

These changes improve the flexibility and accuracy of quality assessments within the application.

### CODE REVIEW

File: changelog.d/20241101_140759_mzhiltso_compare_point_groups_in_image_space.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+### Added
+
+- A quality setting to compare point groups without using bbox
+  (<https://github.com/cvat-ai/cvat/pull/8634>)
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply changelog.d/20241101_140759_mzhiltso_compare_point_groups_in_image_space.md:4]: This is added in 20241106_170626_mzhiltso_match_empty_frames.md

File: cvat/apps/quality_control/migrations/0004_qualitysettings_use_bbox_size_for_points.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="qualitysettings",
+            name="use_bbox_size_for_points",
+            field=models.BooleanField(default=True),
+            preserve_default=False,
+        ),
+    ]
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/quality_control/migrations/0004_qualitysettings_use_bbox_size_for_points.py]: So, after the existing records get an initial value during migration, they will have the value set, aren't they? New records will have the value initialized on the server side.

File: cvat/apps/quality_control/models.py [hunk 3]
```diff
@@ -205,6 +205,8 @@ class QualitySettings(models.Model):
 
     low_overlap_threshold = models.FloatField()
 
+    use_bbox_size_for_points = models.BooleanField()
+
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/quality_control/models.py]: The default value for the field is provided in the model constructor.

File: tests/python/shared/assets/quality_settings.json [hunk 4]
```diff
@@ -14,13 +14,15 @@
       "line_orientation_threshold": 0.1,
       "line_thickness": 0.01,
       "low_overlap_threshold": 0.8,
+      "match_empty_frames": false,
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply tests/python/shared/assets/quality_settings.json:17]: The option is checked in one of the `test_settings_affect_metrics` test parametrizations.

File: cvat/apps/quality_control/quality_reports.py [hunk 5]
```diff
@@ -214,6 +217,13 @@
     panoptic_comparison: bool = True
     "Use only the visible part of the masks and polygons in comparisons"
 
+    match_empty_frames: bool = False
+    """
+    Consider unannotated (empty) frames as matching. If disabled, quality metrics, such as accuracy,
+    will be 0 if both GT and DS frames have no annotations. When enabled, they will be 1 instead.
+    This will also add virtual annotations to empty frames in the comparison results.
+    """
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py:225]: The existing code already initializes all the known values, including this one.

File: cvat/apps/quality_control/quality_reports.py [hunk 6]
```diff
@@ -1525,6 +1544,7 @@
             panoptic_comparison=settings.panoptic_comparison,
             iou_threshold=settings.iou_threshold,
             oks_sigma=settings.oks_sigma,
+            use_bbox_size_for_points=settings.use_bbox_size_for_points,
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py]: There is a comment about this in the line 1551, it's handled differently.

File: cvat-core/src/quality-settings.ts [hunk 7]
```diff
@@ -197,6 +217,7 @@ export default class QualitySettings {
         const result: SerializedQualitySettingsData = {
             iou_threshold: this.#iouThreshold,
             oks_sigma: this.#oksSigma,
+            use_bbox_size_for_points: this.#useBboxSizeForPoints,
```
[the remark this answers is not in the document]
Comment (zhiltsov-max) [reply cvat-core/src/quality-settings.ts]: The schema is updated.

File: cvat/apps/quality_control/quality_reports.py [hunk 8]
```diff
@@ -1972,8 +1979,16 @@ def _find_closest_unmatched_shape(shape: dm.Annotation):
             gt_label_idx = label_id_map[gt_ann.label] if gt_ann else self._UNMATCHED_IDX
             confusion_matrix[ds_label_idx, gt_label_idx] += 1
 
+        if self.settings.match_empty_frames and not gt_item.annotations and not ds_item.annotations:
+            # Add virtual annotations for empty frames
```
Comment (SpecLad) [inline cvat/apps/quality_control/quality_reports.py:1983]: What's the purpose of these virtual annotations?

Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py:1983]: Basically, to break math and code less and get the expected values from `accuracy`, `precision`, and `recall` here and in aggregated reports (job report, task report).

File: cvat/apps/quality_control/quality_reports.py [hunk 9]
```diff
@@ -2015,9 +2030,8 @@ def _make_zero_confusion_matrix(self) -> Tuple[List[str], np.ndarray, Dict[int,
 
         return label_names, confusion_matrix, label_id_idx_map
 
-    @classmethod
-    def _generate_annotations_summary(
-        cls, confusion_matrix: np.ndarray, confusion_matrix_labels: List[str]
+    def _compute_annotation_summary(
```
Comment (SpecLad) [inline cvat/apps/quality_control/quality_reports.py:2033]:
```suggestion
    def _compute_annotations_summary(
```

This would be consistent with the result type name.

File: cvat/apps/quality_control/quality_reports.py [hunk 10]
```diff
@@ -2082,19 +2108,52 @@ def generate_report(self) -> ComparisonReport:
...
+        if self.settings.match_empty_frames and empty_frame_count:
+            # Add virtual annotations for empty frames,
+            # they are not included in the confusion matrix
+            annotation_summary.valid_count += empty_frame_count
+            annotation_summary.total_count += empty_frame_count
+            annotation_summary.ds_count += empty_frame_count
+            annotation_summary.gt_count += empty_frame_count
+
+        # Cannot be computed in accumulate()
+        annotation_components.shape.mean_iou = np.mean(mean_ious)
```
Comment (SpecLad) [inline cvat/apps/quality_control/quality_reports.py:2140]: This doesn't seem right. A mean of means is not necessarily the same as the mean of all underlying values.

(It also seems unrelated to the topic of the PR.)

File: cvat/apps/quality_control/quality_reports.py [hunk 11]
```diff
@@ -2082,19 +2108,52 @@ def generate_report(self) -> ComparisonReport:
...
+            confusion_matrix, confusion_matrix_labels
+        )
+
+        if self.settings.match_empty_frames and empty_frame_count:
+            # Add virtual annotations for empty frames,
+            # they are not included in the confusion matrix
+            annotation_summary.valid_count += empty_frame_count
+            annotation_summary.total_count += empty_frame_count
+            annotation_summary.ds_count += empty_frame_count
+            annotation_summary.gt_count += empty_frame_count
```
Comment (SpecLad) [inline cvat/apps/quality_control/quality_reports.py:2137]: I'm not sure if this is really a useful thing to do here. Imagine a situation where there are 100 frames, frame 1 has 1 valid annotations and 2 total annotations, every other frame is empty.

In this case, with `match_empty_frames` off:

* frame 1 has accuracy 50%, others have 0%.
* total accuracy is 50%.

With `match_empty_frames` on:

* frame 1 has accuracy 50%, others have 100%.
* total accuracy is (1+99)/(2+99) = 99%.

Do you think this jump in total accuracy would be expected by the user?

File: cvat/apps/quality_control/quality_reports.py (hunk 8 above)
Comment (SpecLad) [reply cvat/apps/quality_control/quality_reports.py:1983]: I see, but it seems like bad UX. You choose to match empty frames, and you inexplicably get fake totals. Moreover, if I understand the code correctly, with this change the totals no longer match the confusion matrix, which seems like it could cause more confusion and math errors down the line.

Is this solution really that much less disruptive than fudging the metric formulas to return 1 instead of 0?

> It's not totally nonsense, as an empty annotation can be considered a frame annotation by itself.

This is a valid point, but you're not implementing that consistently. In your implementation, this "empty annotation" only appears when both GT and DS frames are empty. To do this consistently, you'd need to increase the GT count whenever the GT frame is empty, and the DS count whenever the DS frame is empty (and the valid count when both are).

You could also resolve the inconsistency between the confusion matrix and the totals by adding another row/column to the matrix specifically for these "empty annotations".

This would resolve the consistency issues, and slightly improve the UX issue, since you could see in the report where the extra annotation is coming from. But frankly, it still seems easier to me to just fudge the metrics.

File: cvat/apps/quality_control/quality_reports.py (hunk 10 above)
Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py:2140]: It's how it worked before this PR, I didn't plan to change anything here. Also, we don't have too many options here on how to compute the value.

File: cvat/apps/quality_control/quality_reports.py (hunk 11 above)
Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py:2137]: It's hard to say about the expectations, to me the user just works in one of the modes - they either consider empty frames as annotated or not. But the second value seems to be more correct, so maybe something should be done for the case with the option disabled.

File: cvat/apps/quality_control/quality_reports.py [hunk 12]
```diff
@@ -1972,8 +1979,16 @@ def _find_closest_unmatched_shape(shape: dm.Annotation):
             gt_label_idx = label_id_map[gt_ann.label] if gt_ann else self._UNMATCHED_IDX
             confusion_matrix[ds_label_idx, gt_label_idx] += 1
 
+        if self.settings.match_empty_frames and not gt_item.annotations and not ds_item.annotations:
+            # Add virtual annotations for empty frames
+            valid_labels_count = 1
+            total_labels_count = 1
+
+            valid_shapes_count = 1
+            total_shapes_count = 1
```
Comment (SpecLad) [inline cvat/apps/quality_control/quality_reports.py:1988]: Shouldn't `ds_shapes_count` and `gt_shapes_count` be updated too?

File: cvat/apps/quality_control/quality_reports.py (hunk 10 above)
Comment (SpecLad) [reply cvat/apps/quality_control/quality_reports.py:2140]: Ah, okay, I didn't see it was moved from a different section.

> Also, we don't have too many options here on how to compute the value.

FWIW, I would imagine you could change the `mean_iou` member of `ComparisonReportAnnotationShapeSummary` to `sum_iou`, and compute the mean on-demand instead.

File: cvat/apps/quality_control/quality_reports.py (hunk 12 above)
Comment (zhiltsov-max) [reply cvat/apps/quality_control/quality_reports.py:1988]: Thanks, updated. I guess the idea was that these are not shapes exactly, but if we add precision and recall in the shape summary section, the metrics will be broken.

**Review by SpecLad — APPROVED** [review]
